pyro/infer/mcmc/hmc.py

```diff
@@ -227,102 +209,36 @@ def _find_reasonable_step_size(self, z):
-        if self._adapt_window_ending == self._warmup_steps:
-            self._end_warmup()
-            return
-
-        if self._adapt_window_ending == self._adapt_start_buffer:
-            if self.adapt_mass_matrix:
-                self._adapt_mass_matrix_phase = True
-        else:
-            if self.adapt_step_size:
-                self._step_size_adapt_scheme.prox_center = math.log(10 * self.step_size)
-                self._step_size_adapt_scheme.reset()
-
-            if self.adapt_mass_matrix:
-                self._inverse_mass_matrix = self._mass_matrix_adapt_scheme.get_covariance()
-                self._update_r_dist()
-                self._mass_matrix_adapt_scheme.reset()
-
-            self._adapt_window = 2 * self._adapt_window
-            if self._adapt_window_ending == self._adapt_mass_matrix_phase_ending:
-                self._adapt_mass_matrix_phase = False
-                self._adapt_window_ending = self._warmup_steps
-                return
+                initial_step_size = self._find_reasonable_step_size(z)
 
-        self._adapt_window_ending = self._adapt_window_ending + self._adapt_window
-        # expanding the current window if length of the next one is too large
-        next_adapt_window_ending = self._adapt_window_ending + 2 * self._adapt_window
-        if next_adapt_window_ending > self._adapt_mass_matrix_phase_ending:
-            self._adapt_window_ending = self._adapt_mass_matrix_phase_ending
-
-    def _adapt_step_size(self, accept_prob):
-        # calculate a statistic for Dual Averaging scheme
-        H = self._target_accept_prob - accept_prob
-        self._step_size_adapt_scheme.step(H)
-        log_step_size, _ = self._step_size_adapt_scheme.get_state()
-        self.step_size = math.exp(log_step_size)
-        self.num_steps = max(1, int(self.trajectory_length / self.step_size))
-
-    def _update_r_dist(self):
-        loc = self._inverse_mass_matrix.new_zeros(self._inverse_mass_matrix.size(0))
-        if self.full_mass:
-            self._r_dist = dist.MultivariateNormal(loc,
-                                                   precision_matrix=self._inverse_mass_matrix)
-        else:
-            self._r_dist = dist.Normal(loc, self._inverse_mass_matrix.rsqrt())
+        self._adapter.configure(self._warmup_steps,
+                                self._initial_mass_matrix,
+                                initial_step_size)
 
     def _sample_r(self, name):
-        r_flat = pyro.sample(name, self._r_dist)
+        r_dist = self._adapter.r_dist
+        # For finding a reasonable step size before adapter is initialized.
+        if r_dist is None:
+            inv_mass_matrix = self._initial_mass_matrix
+            loc = inv_mass_matrix.new_zeros(inv_mass_matrix.size(0))
+            if self.full_mass:
+                r_dist = dist.MultivariateNormal(loc, precision_matrix=inv_mass_matrix)
+            else:
+                r_dist = dist.Normal(loc, inv_mass_matrix.rsqrt())
```

Comment:
I think that we can move this logic to Adapter too.

Reply:
`r_dist` is already moved to adapter. This is one part that I wish could be more cleanly separted - the issue here is that some components of the adapter (those that rely on the model trace like `inverse_mass_matrix` etc.) are not initialized until we hit `setup._configure_adaptation` but that method itself requires an `initial_step_size` for which we need to hit `find_reasonable_step_size`. None of the alternatives seemed cleaner :( - one of which would be to initialize the adapter in three stages: constructing it, then initializing the mass matrix and then finally setting the initial step size.

Reply:
Btw @fehiepsi - Feel free to directly commit to this PR if you notice something off!

Reply:
I think that you can revise the logic of `adapter.configure` like:
```
if warmup_steps is not None:
    update schedule
if step_size is not None:
    update step_size
if inv_mass_matrix is not None:
    update inv_mass_matrix
```
+ then remove `_configure_adaptation`, putting the following logic 
```
self._adapter.configure(warmup_steps=warmup_steps, inv_mass_matrix=initial_mass_matrix)
if self.adapt_step_size:
    with pyro.validation_enabled(False):
          initial_step_size = self._find_reasonable_step_size(trace)  # trace instead of z
    self._adapter.configure(step_size=initial_step_size)
```
at the end of `setup`.

This way, you can remove `self._initial_mass_matrix` in HMC I guess.

Reply:
I think at this point, let us just have separate setters - and then assert that we have a well specified configuration inside `configure` method. The issue with multiple optional parameters is that we won't be able to verify that the configuration is incomplete. Will make the change.